### PR TITLE
[Feature](bangc-ops): add indice convolution backward filter op

### PR DESCRIPTION
**Pull request by alex-xuwenming**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

Please describe your motivation and the goal you want to achieve through this pull request.

## 2. Modification

Please briefly describe what modification is made in this pull request, and indicate where to make the modification.

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](../docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- [x] (half)diff1, diff1 <= 3e-3
- [x] (float)diff2, diff2 <= 1e-5

#### 3.1.2 Operator Scheme checklist

|     No.        |                 Details              |            Check Results             |
|----------------|--------------------------------------|--------------------------------------|
|        1       |Supported hardware                    |             MLU370<br>MLU590         |
|        2       |Job types                             |          host 拼接实现      |
|        3       |Layouts                               |          ARRAY    |
|        4       |Whether multi-dimensions are supported|            支持                          |
|        5       |Whether element zero is supported     |       支持（跳过计算）                               |
|        6       |Data type(half/float)                 |           half / float          |
|        7       |Whether there is size limit           |               单唯度不超过int32                    |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test
- [x] Multidimensional tensor test
- [x] Layout test
- [x] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see[GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](../docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

|                   Test Point                    | Acceptance Standard | Test Result (Error Message) |
| ----------------------------------------------- | --------------------| --------------------------- |
| Whether it conforms to the operator restriction |     Normal error    |                             |
| Whether illegal parameters are passed           |     Normal error    |                             |

### 3.2 Accuracy Test

For the cases used in the New Feature Test section, the features and the number of cases are recorded here. When multiple operations are tested, multiple tables are needed to include details of these operations.

Operation:

|Test Point           | Description                      | Quantity |  Comment |
|----------           |----------------------------------|----------|  --------|
|Data type test       |half/float                  |       |       all pass   |
|Mult-tensor test     |conv2D, conv3D      |           |    all pass     |
|Layout test          |Supports Array                |           |    all pass      |
|Zero element test    |support      |          |     all pass     |
|Stability test       |--gtest_repeat=100<br>--thread=100|          |   all pass       |
|Mult-platform test   |MLU370/MLU590                     |           |    all pass      |
|Nan/INF test         | depend by matmul    |          |          |

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](../docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|-----|----|----|----|-----|
|indice_convolution_backward_filtr|  / | /   |  /   |  /  |  /  |  /  |  /   |

Platform：MLU590

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|----|----|----|----|-----|
|indice_convolution_backward_filter|  1.9ms  |  /  |  /  | /   |   / |   float32 |  features[58838, 128], filters[3,3,3,128,128], output[58838, 128]   |

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: docs/bangc-docs/design_docs/indice_convolution/indice_convolution_backward_filter.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
+# roi_crop 算子开发设计方案
```
[the remark this answers is not in the document]
Comment (alex-xuwenming) [reply docs/bangc-docs/design_docs/indice_convolution/indice_convolution_backward_filter.md]: 已修改

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -7187,6 +7187,161 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+                                                                                 const mluOpTensorDescriptor_t features_desc,
+                                                                                 const mluOpTensorDescriptor_t output_grad_desc,
+                                                                                 const mluOpTensorDescriptor_t indice_pairs_desc,
+                                                                                 const mluOpTensorDescriptor_t filters_grad_desc,
+                                                                                 const int64_t indice_num[],
+                                                                                 const int64_t inverse,
+                                                                                 const int64_t sub_m,
+                                                                                 size_t *workspace_size);
+
+// Group:MatMul
```
[the remark this answers is not in the document]
Comment (alex-xuwenming) [reply bangc-ops/mlu_op.h]: 已修改

**Review by Devin-D-U — APPROVED** [review]

File: bangc-ops/core/logging.h [hunk 3]
```diff
@@ -111,7 +111,7 @@
     if ((__status__) != MLUOP_STATUS_SUCCESS) {                            \
       LOG(ERROR) << api << "BAD return status: " << #status << " returns " \
                  << (__status__) << " (FILE: " << __FILE__                 \
-                 << ", LINE: " << __LINE__ "). " __VA_ARGS__;              \
+                 << ", LINE: " << __LINE__  << "). " __VA_ARGS__;          \
```
[the remark this answers is not in the document]
Comment (alex-xuwenming) [reply bangc-ops/core/logging.h:114]: 会出错，之前库里没人使用，所以没报出来。

File: bangc-ops/kernels/indice_convolution_backward_filter/indice_convolution_backward_filter.cpp [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+}
+
+static void indiceConvFilterGencase(
+    mluOpHandle_t handle, const mluOpTensorDescriptor_t features_desc,
+    const void *features, const mluOpTensorDescriptor_t output_grad_desc,
+    const void *output_grad, const mluOpTensorDescriptor_t indice_pairs_desc,
+    const void *indice_pairs, const int64_t indice_num[], const int64_t inverse,
+    const int64_t subm, void *workspace, size_t workspace_size,
+    const mluOpTensorDescriptor_t filters_grad_desc, void *filters_grad) {
+  double indice_num_max =
```
Comment (Unireverse) [inline bangc-ops/kernels/indice_convolution_backward_filter/indice_convolution_backward_filter.cpp:83]: 这个变量看起来没被用到

File: docs/bangc-docs/design_docs/indice_convolution/indice_convolution_backward_filter.md [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+### 1.1 算子需求分析
+
+该需求分析为框架原生算子实现功能的需求分析，对于框架原生支持但 MLU-OPS 当前版本不支持的功能，需要在`1.4算子限制` 章节中显式注明。未明确注明不支持的功能，默认 MLU-OPS 全部支持。
+
+| 算子功能简介                | 根据感兴趣区域提取固定大小的输出特征|
+| -------------------------- | ---------------------------------------- |
+| 需求来源                    | PyTorch                                 |
+| 应用网络                    | CenterPoint                             |
+| 输入数据类型                | features and output_grad:half/float, indice_pairs: int32 |
+| 输入 Shape                  |1. input:<br>input: [in_active_num,ci]; output_grad: [out_active_num, co]; indice_pairs: [k, 2, in_active_num]<br> output:<br><br>|
```
Comment (WangYuan-cn) [inline docs/bangc-docs/design_docs/indice_convolution/indice_convolution_backward_filter.md:42]: output:<br><br>  缺少内容
建议 都加序号换行，或把`1. `删掉

File: bangc-ops/kernels/indice_convolution_backward_filter/indice_convolution_backward_filter.cpp [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+               << mluop::getNameOfDataType(pairs_dtype) << ".";
+    return MLUOP_STATUS_BAD_PARAM;
+  }
+
+  if (input_dtype != diffy_dtype || input_dtype != filters_grad_dtype ||
+      !isFloatDtype(input_dtype) || !isFloatDtype(diffy_dtype) ||
+      !isFloatDtype(filters_grad_dtype)) {
+    LOG(ERROR)
+        << api_name << " The data type of features_desc, output_grad_desc "
+        << "and filters_grad_desc should be same and both are half or float."
```
Comment (Unireverse) [inline bangc-ops/kernels/indice_convolution_backward_filter/indice_convolution_backward_filter.cpp:124]:
```suggestion
        << "and filters_grad_desc should be the same and the three should be either half or float."
```

File: bangc-ops/kernels/indice_convolution_backward_filter/indice_convolution_backward_filter.cpp [hunk 7]
```diff
@@ -0,0 +1,557 @@
...
+      (filters_grad_dtype == MLUOP_DTYPE_FLOAT &&
+       filters_grad_on_dtype == MLUOP_DTYPE_HALF)) {
+    LOG(ERROR)
+        << api_name
+        << " The on-chip data type of filters_grad_desc may not be set, "
+        << "if it is set, only half or float types are supported, "
+        << "and the bit width of on-chip data type can not be smaller than "
+        << "that of off-chip data type. But now two data types of "
+           "filters_grad_desc "
+           "are "
```
Comment (Unireverse) [inline bangc-ops/kernels/indice_convolution_backward_filter/indice_convolution_backward_filter.cpp:169]: 这个是不是可以移动到上面一行

File: bangc-ops/kernels/indice_convolution_backward_filter/indice_convolution_backward_filter.cpp [hunk 8]
```diff
@@ -0,0 +1,557 @@
...
+  //       MLUOP_LAYOUT_NDHWC == filters_grad_layout)) {
+  //   LOG(ERROR) << api_name << " The layout of filters_grad_desc is "
+  //              << mluop::getNameOfTensorLayout(filters_grad_layout)
+  //              << ", which is unsupported.";
+  //   return MLUOP_STATUS_BAD_PARAM;
+  // }
+  // auto ci = mluOpGetTensordimC(filters_grad_desc);
+  // auto co = mluOpGetTensordimN(filters_grad_desc);
+  // int32_t kd = filters_grad_desc->dim == 4 ? 1 :
+  // mluOpGetTensordimD(filters_grad_desc);
```
Comment (Unireverse) [inline bangc-ops/kernels/indice_convolution_backward_filter/indice_convolution_backward_filter.cpp:221]: 这一段还要吗

File: docs/bangc-docs/design_docs/indice_convolution/indice_convolution_backward_filter.md [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+| handle            |MLU-OPS 上下文的指针            |mluOpHandle_t| /        | 无       |
+| features_desc     |卷积输入，输入的有效点数据      |half/float   | ARRAY    | 无       |
+| output_grad_desc  |卷积输出的梯度，输出有效点的梯度|half/float   | ARRAY    | 无       |
+| indice_pairs_desc |输入，输出和卷积核的映射关系，  |int32        | ARRAY    | 无       |
+| indice_pairs_num  |卷积核中的每个点进行计算的次数  |int64        | ARRAY    | 无       |
+| filters_grad_desc |卷积权值的梯度                  |half/float   | ARRAY    | 无       |
+| sub_m             |稀疏卷积的模式                  |int64        | /        | 无       |
+| inverse           |稀疏卷积的算法，目前不支持      |int64        | /        | 无       |
+
+- 当前限制
```
Comment (WangYuan-cn) [inline docs/bangc-docs/design_docs/indice_convolution/indice_convolution_backward_filter.md:81]: 缺少 算子限制 小节，建议按算子模板修改小节名称， https://github.com/Cambricon/mlu-ops/blob/master/docs/bangc-docs/BANGC-OPS-Operator-Design-Doc-Template.md

File: bangc-ops/mlu_op.h [hunk 10]
```diff
@@ -7187,6 +7187,161 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
                                 const mluOpTensorDescriptor_t output_desc,
                                 void *output);
 
+
+// Group:indiceConvolution
+/*!
+ * @brief Returns in \b workspace_size the size of the MLU memory that is used as an extra workspace
+ * to optimize the indice_convolution_backward_filter operation.
```
Comment (Unireverse) [inline bangc-ops/mlu_op.h:7194]: 这里的 indice_convolution_backward_filter 改成 ::mluOpIndiceConvolutionBackwardFilter 看起来比较好

**Review by Unireverse — APPROVED** [review]

File: bangc-ops/kernels/indice_convolution_backward_filter/indice_convolution_backward_filter.cpp [hunk 11]
```diff
@@ -0,0 +1,557 @@
...
+  CHECK_RETURN(api_name,
+               mluOpSetMatMulDescAttr(matmul_desc, MLUOP_MATMUL_ALLOW_TF32,
+                                      &(allow_tf32), sizeof(int32_t)));
+  return MLUOP_STATUS_SUCCESS;
+}
+
+inline std::string getTensorShapeString(const mluOpTensorDescriptor_t desc) {
+  std::string res;
+  res.push_back('[');
+  for (int i = 0; i < desc->dim - 1; i++) {
```
Comment (WangYuan-cn) [inline bangc-ops/kernels/indice_convolution_backward_filter/indice_convolution_backward_filter.cpp:69]: 同一文件，建议统一类型名称的使用， 56 行 int32_t

File: bangc-ops/kernels/indice_convolution_backward_filter/indice_convolution_backward_filter.cpp [hunk 12]
```diff
@@ -0,0 +1,557 @@
...
+
+  CHECK_RETURN(api_name, mluOpDestroyTensorDescriptor(trans_in_desc));
+  CHECK_RETURN(api_name, mluOpDestroyTensorDescriptor(trans_out_desc));
+  CHECK_RETURN(api_name, mluOpDestroyTransposeDescriptor(trans_desc));
+  return MLUOP_STATUS_SUCCESS;
+}
+
+// called by getWorkspace and compute api
+// workspace_size is not nullptr when it's from getWorkspace api.
+mluOpStatus_t internalIndiceConvBackwardFilter(
```
Comment (WangYuan-cn) [inline bangc-ops/kernels/indice_convolution_backward_filter/indice_convolution_backward_filter.cpp:311]: 建议修改为 static

File: bangc-ops/mlu_op.h [hunk 13]
```diff
@@ -7187,6 +7187,161 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ *
+ * The size of extra workspace is based on the given information of the indice_convolution_backward_filter
+ * operation, including the input tensor descriptor \b features_desc, \b output_grad_desc and \b indice_pairs_desc, output
+ * tensor descriptor \b filters_grad_desc, and the array \b indice_num[]. 
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and queues in the
+ * indice_convolution_backward_filter operation. For detailed information, see ::mluOpHandle_t.
+ * @param[in] features_desc
+ * The descriptor of features that need to convolution. For detailed information,
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7204]: 去掉to

File: bangc-ops/mlu_op.h [hunk 14]
```diff
@@ -7187,6 +7187,161 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * see ::mluOpTensorDescriptor_t.
+ * @param[in] output_grad_desc
+ * The descriptor of output grad. For detailed information,
+ * see ::mluOpTensorDescriptor_t.
+ * @param[in] indice_pairs_desc
+ * The descriptor of indice pairs between inputs locations and outputs locations. For detailed information,
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7210]: inputs locations and outputs locations > input locations and output location

File: bangc-ops/mlu_op.h [hunk 15]
```diff
@@ -7187,6 +7187,161 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * see ::mluOpTensorDescriptor_t.
+ * @param[in] filters_grad_desc
+ * The descriptor of filters grad tensor. For detailed information,
+ * see ::mluOpTensorDescriptor_t.
+ * @param[in] indice_num
+ * Pointer to the host memory that stores the indice pairs num.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7216]: num > number

File: bangc-ops/mlu_op.h [hunk 16]
```diff
@@ -7187,6 +7187,161 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * @param[in] inverse
+ * Currently not supported and should be set to 0.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7218]: Currently it is not supported and should be set to 0.

File: bangc-ops/mlu_op.h [hunk 17]
```diff
@@ -7187,6 +7187,161 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * @param[in] sub_m
+ * Currently not supported and should be set to 0.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7220]: Currently it is not supported and should be set to 0.

File: bangc-ops/mlu_op.h [hunk 18]
```diff
@@ -7187,6 +7187,161 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+// Group:indiceConvolution
+/*!
+ * @brief Computes the indice_convolution_backward_filter operation, then returns the results in the output
+ * tensor \b filters_grad.
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and queues in the
+ * indice_convolution_backward_filter operation. For detailed information, see ::mluOpHandle_t.
+ * @param[in] features_desc
+ * The descriptor of features that need to convolution. For detailed information,
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7262]: 去掉to

File: bangc-ops/mlu_op.h [hunk 19]
```diff
@@ -7187,6 +7187,161 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * Pointer to the MLU memory that stores the output grad tensor.
+ * @param[in] indice_pairs_desc
+ * The descriptor of indice pairs between inputs locations and outputs locations. For detailed information,
+ * see ::mluOpTensorDescriptor_t.
+ * @param[in] indice_pairs
+ * Pointer to the MLU memory that stores the indice pairs tensor.
+ * @param[in] indice_num
+ * Pointer to the host memory that stores the indice pairs num.
+ * @param[in] inverse
+ * Currently not supported and should be set to 0.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7279]: Currently it is not supported and should be set to 0.

File: bangc-ops/mlu_op.h [hunk 20]
```diff
@@ -7187,6 +7187,161 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * @param[in] sub_m
+ * Currently not supported and should be set to 0.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7281]: Currently it is not supported and should be set to 0.

File: bangc-ops/mlu_op.h [hunk 21]
```diff
@@ -7187,6 +7187,161 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM, ::MLUOP_STATUS_ARCH_MISMATCH
+ *
+ * @par Data Type
+ * - This function supports the combinations of the following data types for
+ *   input tensor \b features, \b output_grad, \b indice_pairs_num and output tensor \b filters_grad.
+ *   - \b features, \b output_grad, \b indice_pairs \b filters_grad data type: half, half, int32, half
+ *   - \b features, \b output_grad, \b indice_pairs \b filters_grad data type: float, float, int32, float
+ *
+ * @note
+ * - This function only supports on MLU300 series or above platforms.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7303]: This function is only supported on MLU300 series or above platforms.

File: bangc-ops/mlu_op.h [hunk 22]
```diff
@@ -7187,6 +7187,161 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * - This function does not support set tensor onchip data type with fixed-point type.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7304]: setting

File: bangc-ops/mlu_op.h [hunk 23]
```diff
@@ -7187,6 +7187,161 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ *
+ * @par Scale Limitation
+ * - The input tensors and output tensor must meet the following requirements:
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7307]: tensor

File: bangc-ops/mlu_op.h [hunk 24]
```diff
@@ -7187,6 +7187,161 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ *   - The \b features and \b output_grad must be two dimensions.
+ *   - The \b indice_pairs must be three dimensions, and the first dimension value must be euqal to kernel size of \b filters_grad,
+ *     the second dimension must be 2, and the last dimension must be the same as the number of \b features first dimension.
+ *   - The \b filters_grad should be four or five dimensions. The last dimension of \b filters_grad must be the same as the last dimension of \b output_grad,
+ *     and the penultimate dimension of \b filters_grad must be same as the last dimension of \b features.
+ *   - The array length of indice_num must be same as the first dimension of \b indice_pairs.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7313]: the same

**Review by GraceGuanH — COMMENTED** [review]
Please check the comments.

File: bangc-ops/kernels/indice_convolution_backward_filter/indice_convolution_backward_filter.cpp (hunk 4 above)
Comment (alex-xuwenming) [reply bangc-ops/kernels/indice_convolution_backward_filter/indice_convolution_backward_filter.cpp]: 已删除

File: docs/bangc-docs/design_docs/indice_convolution/indice_convolution_backward_filter.md (hunk 5 above)
Comment (alex-xuwenming) [reply docs/bangc-docs/design_docs/indice_convolution/indice_convolution_backward_filter.md]: 已经修改

File: bangc-ops/mlu_op.h (hunk 17 above)
Comment (alex-xuwenming) [reply bangc-ops/mlu_op.h]: 这里写错了，应该是支持0或者非0的，如果是非0就代表是sub_m模式的卷积。改了写法，你在检查一下吧。

File: bangc-ops/mlu_op.h (hunk 20 above)
Comment (alex-xuwenming) [reply bangc-ops/mlu_op.h]: 同getWorkspace接口，换了写法

**Review by WangYuan-cn — APPROVED** [review]

**Review by GraceGuanH — APPROVED** [review]

**Pull request merged**